### PR TITLE
[WIP] python3 transition

### PR DESCRIPTION
**Pull request by trws**
See the commit messages for some more detail, but this PR makes python 3.6 required to build flux and adds a new subpackage for python2 bindings that is built if the `--with-python2` flag is passed to configure.  There is a bit more to do here, mainly hooking in with travis and figuring out what we want to do with/about pythonpath, but I wanted to get this up here so people have a chance to look at it before it gets much further along.

### CODE REVIEW

File: src/bindings/python-2/configure.ac [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+AM_CHECK_PYMOD(yaml,
+               [StrictVersion(yaml.__version__) >= StrictVersion ('3.10.0')],
+               ,
+               [AC_MSG_ERROR([could not find python module yaml, version 3.10+ required])]
+               )
+AM_CHECK_PYMOD(jsonschema,
+               [StrictVersion(jsonschema.__version__) >= StrictVersion ('2.3.0')],
+               ,
+               [AC_MSG_ERROR([could not find python module jsonschema, version 2.3.0+ required])]
+               )
```
Comment (SteVwonder) [inline src/bindings/python-2/configure.ac]: I'm not 100% sure on this, but I think this dependency only exists for commands in `src/cmd` and `t/`, which I think will only be run under python3.  So we may be able to drop this dependency for the bindings generated against python2.

Comment (trws) [reply src/bindings/python-2/configure.ac]: Thanks for that, currently this runs all, the python tests from t/, I'll double-check that none of them need it and if not drop it.

**Comment by SteVwonder** [comment]
To summarize and continue our face-to-face conversation, @trws, this changes re-opens the problem in #1999; that is, how do we make it so that `/usr/bin/flux` plays nicely with "non-system" python interpreters.  With this PR, as is, if a user runs `flux subcommand`, `flux` will prepend `$FLUX_PREFIX/lib/python3.6/site-packages` to `PYTHONPATH`, which could "pollute" the `PYTHONPATH` for users that leverage spack, virtualenv, conda, the system-installed python2.7, etc (basically any interpreter other than `/usr/bin/python3.6`).

One solution is to add an argument to `flux` that will preserve an environment variable, which could be useful for scenarios where the user does not want flux to modify `LD_LIBRARY_PATH` or other environment variables.  Strawman example: `flux --preserve-env=PYTHONPATH subcommand ...`

**Comment by trws** [comment]
@garlick, @SteVwonder, this version now appears to be working with dist and everything else.  It's the minimum delta from where it was, and configures the python-2 bindings unconditionally while only building or testing them when enabled.

**Comment by SteVwonder** [comment]
Looks like maybe two builders failed due to a hang. Going to try restarting just those two.

`ERROR: t1003-kvs-stress.t - exited with status 137 (terminated by signal 9?)`

**Comment by garlick** [comment]
Testing with flux-core configured with `--prefix=/usr/ --with-python2` on a Ubuntu 18.04 system:   for some reason, python2 installs to `/usr/lib/python2.7/dist_packages` while python3 installs to `/usr/lib/python3.6/site_packages`.  If I take one of our test programs, add a python2.7 shebang, and try to run it outside of a flux environment (e.g. not in an instance or behind the flux command front end) it finds our modules.   If I do the same with a python3.6 shebang, it fails with
```
Traceback (most recent call last):
  File "./submit-waitany.py", line 18, in <module>
    import flux
ModuleNotFoundError: No module named 'flux'
```

**Comment by trws** [comment]
The placement sounds alright, but I'm now looking at a ubuntu 18.04 docker image and it looks like they are using `/usr/lib/python3.6/dist-packages` and `/usr/lib/python2.7/dist-packages` rather than site-packages, which is rather peculiar in my experience but I'll try to see why the installation path is getting detected incorrectly.

**Comment by garlick** [comment]
A debian thing apparently?

https://stackoverflow.com/questions/9387928/whats-the-difference-between-dist-packages-and-site-packages

On master, `configure --prefix=/usr` (python2) uses  `dist-packages`.
However, `PYTHON_VERSION=3 configure --prefix=usr` uses  `site-packages`.

**Comment by trws** [comment]
So it seems, I think we're running afoul of debian distros making an explicit distinction in their python config between module and extension library install locations.  I had been targeting the install to the pyexecdir, which is the extension library location, but this tells me the site-packages directory (which ironically is dist-packages in this instance) is what we need.  Let me test this real quick but I think I've got it.

**Comment by trws** [comment]
It was nastier than I thought.  We were using the right thing, but AM_PATH_PYTHON doesn't work correctly if certain packages are available and you're using debian's directory structure.  The changes I just uploaded fixed it for me on an ubuntu 19.10 box.

**Comment by garlick** [comment]
Attempting to verify this fix...

I did open a bug on the inconsistency found in current master (your fix can close it if it works, and we have a way to build correct packages on both debian and redhat derived distros).

One minor nit, this cut and paste from sched should probably not be included
```
  flux-python-2 version 0.13.0-474-g467f4d2db
  Prefix...........: /usr
  Debug Build......:
  C Compiler.......: gcc
  C++ Compiler.....:
  CFLAGS...........: -g -O2
  CPPFLAGS..........
  CXXFLAGS..........
  FLUX.............:
  FLUX_CORE_CFLAGS.:
  FLUX_CORE_LIBS...:
  FLUX_PREFIX......:
  LDFLAGS..........:
  LIBS.............:
  Linker...........: /usr/bin/ld -m elf_x86_64
```

**Comment by garlick** [comment]
That worked, sort of - it seems now python3 is being installed into /`usr/lib/python3` instead of `/usr/lib/python3.6`?   There seems to be content in both directories so I'm not sure if it's wrong, but it seems to be different.

**Comment by garlick** [comment]
Incidentally, removal of mrpc just landed on master, so you might rebase and drop mrpc from python2 specific binding code.

Ah so the PYTHON_SITE_DIR you used boils down to asking the installed python what dir to use which sounds right to me, e.g.
```console
$ python3.6 -c "import distutils.sysconfig; print (distutils.sysconfig.get_python_lib(0,0));"
/usr/lib/python3/dist-packages
$ python2.7 -c "import distutils.sysconfig; print (distutils.sysconfig.get_python_lib(0,0));"
/usr/lib/python2.7/dist-packages
```
I would assume that would be distro agnostic, the distro choice having been expressed in how python was originally configured.

**Comment by trws** [comment]
Right, that's exactly what it does and why I thought it made sense.  The problem is that if I test it with python 3.7 and 3.8 on the same install (3.8 is an ubuntu-installed side-install) they both want to put everything in `/usr/lib/python3/dist-packages`.  There has to be a way to get the right thing, but it seems none of the queries used by am_path_python do it, at least not without some outside input I must not be providing.

**Comment by garlick** [comment]
Oh you mean it ignores the configured prefix?   Ah that's a problem.

**Comment by trws** [comment]
I don't believe so, though I'll double check, but the result of that query of sysconfig is incorrect if you have two versions of python3 on one box.

**Comment by trws** [comment]
I'm not sure quite what to do about this.  The script automake uses to detect the appropriate path does the right thing where alternate prefixes are concerned, so that's good.  It does the right thing consistently on python 2.7, but on python3+ on ubuntu things are not good.  The situation seems to be:

1. there is a comment in the script saying they attempt to use `sysconfig` rather than `distutils.sysconfig` for better compatibility with 3
2. if I test the script with `sysconfig` available to it, it says to put things in `/usr/lib/python3.7/site-packages`, which is not part of the default ubuntu sys.path for python3
3. if I make `sysconfig` unavailable, it wants to put them in `/usr/lib/python3/dist-packages`, which doesn't seem to be in sys.path either but somehow it then gets found on import (not sure why that works)

So, using PYTHON_SITE_PKG was definitely the wrong direction, but short of writing some brand new logic here, I don't immediately see a clean way to fix this. There must be a way to get the correct path for pip and so-forth to be able to do it, and the mechanism being used by automake is the one I see recommended, but something is off.

**Comment by trws** [comment]
Thanks to a little gem in a comment [here](https://jasonstitt.com/setuptools-bin-directory) I worked out something that seems to *actually work*.  This update includes a modified python.m4 that uses the distutils install mechanism itself to request the appropriate platform-dependent install location for modules.  It's the only way I've managed to find to get the correct subversion-sensitive path out of the ubuntu built python.  Also did a rebase to remove the erronious intermediate approach.

**Comment by garlick** [comment]
Just built and installed that, and we seem to be back to this (python3 is in site-packages again)
```
$ find /usr/lib/ -name flux
/usr/lib/python3.6/site-packages/flux
/usr/lib/flux
/usr/lib/flux/python3.6/flux
/usr/lib/flux/python2.7/flux
/usr/lib/python2.7/dist-packages/flux
```

**Comment by garlick** [comment]
I think we're out of time on this @trws.  I'll drop the release milestone and we'll shoot for the next one.

**Comment by trws** [comment]
Agreed, it's frustrating but there's a fundamental issue here I can't 
seem to get past.  Will take some more time to mull on it and we can go 
for it next time.

**Comment by SteVwonder** [comment]
Now that #2805 is open can we close this one?

Also, @trws, does #2805 have the dist-packages debian oddity stuff from this PR in it?

**Comment by trws** [comment]
Sure, just don’t delete the branch. 

It does not have the Debian stuff, I’ll check whether it needs it, but it may not since this one is making no attempt to support two installs.

**Pull request closed without merging**